### PR TITLE
std::thread::LocalKeyState: Add state Initializing

### LINKED ISSUE #43491: std::thread::LocalKeyState: Add fourth Initializing state
Right now, [`LocalKeyState`](https://doc.rust-lang.org/std/thread/enum.LocalKeyState.html) contains three states: `Uninitialized`, `Valid`, and `Destroyed`.

Unfortunately, this is not enough to cover all cases. In particular, consider the following code (adapted from some code I wrote as part of writing a global allocator - to motivate that this is a real-world example):

```rust
fn local_or_global(foo: Foo) -> Bar {
    match LOCAL.state() {
        LocalKeyState::Uninitialized | LocalKeyState::Valid => LOCAL.with(|local| do_computation(foo, local)),
        LocalKeyState::Destroyed => do_computation(foo, &GLOBAL),
    }
}
```

The idea here is to ensure that the local key hasn't yet been destroyed - and if it has, to fall back on some global state instead. If the key is valid, then obviously this is fine. If the key is uninitialized, then `with` should perform initialization, which should also be fine...

unless `local_or_global` is called from the initialization routine. This is the case for the allocator code I'm working on since initializing TLS requires performing allocations. In that case, you end up with infinite recursion and eventually a stack overflow.

My proposal is the following: add a fourth `Initializing` state. When `with` is first called and the key needs to be initialized, it is first moved into the `Initializing` state. When initialization completes, it is subsequently moved into the `Valid` state. This allows code that accesses the key to detect whether it's being called from inside the initializer, and behave appropriately in that case (in particular, by _not_ accessing the key and thus recursing the initialization).

The concrete issue I (along with my collaborator on this project, @ezrosent) have run into is that initializing TLS depends on allocation, which depends on TLS, which depends on... In general, this is a problem whenever a TLS key's initialization routine transitively depends on itself (either directly, or through an arbitrarily long cycle of other TLS key dependencies - for example, we've seen similar issues with `Destroyed` keys when using Crossbeam's [epoch-based memory reclamation](https://docs.rs/crossbeam/0.2.10/crossbeam/mem/epoch/index.html), which also uses TLS under the hood).

## Alternatives

It might be tempting to think that just making sure you're in the `Valid` state is a sufficient (if sub-optimal) solution to this problem (and in fact, I thought the same thing for a while). Unfortunately, this solves a chicken-and-egg problem by just deciding that chickens can't exist. By only being willing to access a TLS key in the `Valid` state, you preclude ever calling code that will perform initialization, and thus you'll never move to the `Valid` state in the first place.

### PR DESCRIPTION
**Pull request by joshlf**
Closes #43491.

I know that the issue hasn't been discussed or approved yet, but I figured I'd make this PR to provide something concrete for the discussion.

Open question: This is my first serious PR, and I'm not sure what to do about my `rustfmt` automatically re-formatting existing code that I didn't change. Somebody's guidance here would be appreciated.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks for the PR! Could the formatting related changes be removed from this PR for now?

**Comment by joshlf** [comment]
@alexcrichton I'm not sure how to do that other than to remove them manually - is there a standard `rustfmt` configuration that I can set that will cause `rustfmt` to format code in `libstd` according to the style guidelines (and thus won't make any changes to existing code)? If not, I can just do it manually.

**Comment by alexcrichton** [comment]
@joshlf can you disable auto-rustfmt in your editor, redo the changes, and resubmit them?

**Comment by joshlf** [comment]
@alexcrichton Done.

**Comment by joshlf** [comment]
I think it'd be good to modify `AccessError` to distinguish between its two causes - a key being in the `Initializing` state, and a key being in the `Destroyed` state. Thoughts on what a good API for that would look like? On the one hand, `is_initializing` and/or `is_destroyed` feels a bit unwieldy, but so does adding a new enum just for this (e.g., a `state` method returns `enum AccessErrorState { Initializing, Destroyed }`).

**Comment by joshlf** [comment]
OK, should be ready to go. Removing the WIP tag.

**Comment by arielb1** [comment]
```
[00:45:59] thread 'main' panicked at 'Some tests failed', /checkout/src/tools/compiletest/src/main.rs:323:21
[00:45:59] ------------------------------------------
[00:45:59] error[E0004]: non-exhaustive patterns: `Initializing` not covered
[00:45:59]   --> /checkout/src/test/run-pass/tls-init-on-init.rs:40:19
[00:45:59]    |
[00:45:59] 40 |             match FOO.state() {
[00:45:59]    |                   ^^^^^^^^^^^ pattern `Initializing` not covered
[00:45:59] 
```

**Comment by joshlf** [comment]
@eddyb Can you take a quick look at this and let me know if you think anything here conflicts with [your PR](https://github.com/rust-lang/rust/pull/43746)? I don't think it does, but you know your PR better than I do, and I suspect yours will get merged first so I'll end up rebasing on top of it.

**Comment by eddyb** [comment]
This is a lot of code. But I think all you have to check for if: if you take `&'static self` in `std::thread::local::fast`, you'll have to make that method take `&self` instead (a lifetime error will force you to do so anyway), and mark it `unsafe`.
Then, mark any wrappers created by the macro also `unsafe` if they call one of those `unsafe` methods, and finally, mark the fields in `LocalKey` `unsafe` where necessary.

**Comment by joshlf** [comment]
OK sounds good; thanks!

**Comment by joshlf** [comment]
Looks like [one of the CI tests](https://travis-ci.org/rust-lang/rust/jobs/262481440) failed due to network errors. Does anybody have the ability to tell Travis to retry? Looks like I don't.

```text
remote: Counting objects: 660392, done.
remote: Compressing objects: 100% (123917/123917), done.
error: RPC failed; curl 56 SSLRead() return error -36
fatal: The remote end hung up unexpectedly
fatal: early EOF
fatal: index-pack failed
The command "eval git fetch origin +refs/pull/43550/merge: " failed 3 times.
```

**Comment by arielb1** [comment]
@joshif

Just do a trivial `commit --amend` (changing only the date) and push again. Though now that you have to rebase, it would be better to rebase and that will have another go at travis.

**Comment by joshlf** [comment]
@eddyb Do you have any advice for how to get the lifetimes to work in this PR (in particular, in the `__thread_local_inner` macro)? The issue is that in the original code, there was only a single function that needed to be passed to `LocalKey::new`, so the `#[thread_local] __KEY` could be in the scope of that function's body. In this PR, there are multiple functions, so the key needs to be outside so it can be accessed from all of the functions.

The current version of the code won't compile because `__KEY` does not live long enough for method calls on it to return `'static` references.

**Comment by eddyb** [comment]
@joshlf The one function I had to deal with had the same problem, btw - look at how I changed `fast::Key::get` - specifically, `unsafe` and doesn't take `'static`.

**Comment by eddyb** [comment]
@joshlf Btw you should use `git pull --rebase` to not have merge commits in the PR.

**Comment by joshlf** [comment]
@eddyb Do you know if there's a way to `git pull --rebase` retroactively?

**Comment by eddyb** [comment]
@joshlf You can run it now and it should kill merge commits.

**Comment by joshlf** [comment]
OK it's finally ready for review. Sorry if this isn't the protocol, but I'm new here - pinging @sfackler .

**Comment by alexcrichton** [comment]
Just to make sure I understand this, this is also changing behavior from today, right? Today if you access a TLS key while it's currently being initialized that will succeed, just end up having the value replaced. After this PR, though, it'll be an error? Is that right?

If so, it's technically a regression in behavior but I'm not sure if it'll come up in practice...

In any case, though, would it be possible to implement this perhaps a little less invasively? With the addition of 4 new callbacks it's unclear to me that this won't regress performance (as TLS is pretty performance-sensitive). Would it be possible to just modify the return value of the inner accessor from `Option` to something like `MaybeInitiailizing` or something like that?

**Comment by joshlf** [comment]
> Just to make sure I understand this, this is also changing behavior from today, right? Today if you access a TLS key while it's currently being initialized that will succeed, just end up having the value replaced. After this PR, though, it'll be an error? Is that right?
> 
> If so, it's technically a regression in behavior but I'm not sure if it'll come up in practice...

If for some reason some recursive call to the initializer did not itself recurse (by accessing the variable), then yes, it would have succeeded. However, I expect that there is no existing code that does this - if any recursive accesses happen, then the recursion is unconditional, and so the access would actually result in a stack overflow. In the case of the stack overflow, I don't believe that this represents a regression. However, in the case in which one of the initializer invocations does not recurse, I believe it would technically be a regression, yes.

> In any case, though, would it be possible to implement this perhaps a little less invasively? With the addition of 4 new callbacks it's unclear to me that this won't regress performance (as TLS is pretty performance-sensitive). Would it be possible to just modify the return value of the inner accessor from `Option` to something like `MaybeInitiailizing` or something like that?

I actually architected things this way for performance reasons. A previous version of this code had the `get` callback return a `&LocalKeyState<T>` where `LocalKeyState<T>` is an enum of `Uninitialized`, `Initializing`, `Initialized(T)`, and `Destroyed`. The performance of that implementation was somewhat poorer than that of the nightly branch. The current approach, while more complicated, matches the nightly branch in performance.

The TLDR of the design is that while the overall logic is a bit hairy, the common case is in the fast path, and is quite simple - the `get` callback returns `Some`, and so `try_with` follows its fast path of immediately returning `Ok(f(inner))`.

I definitely admit that the parts besides that - the `pre_init`, `post_init`, and `rollback_init` callbacks - are a bit complicated, but unfortunately they're necessary to allow allocators to use TLS.

**Comment by alexcrichton** [comment]
Do you have some analysis of the performance impact I could poke around with? I'm just wary to store like 10 callbacks per TLS variable...

**Comment by joshlf** [comment]
Sure - I was running the following code to get the performance that I cited in the previous comment:

```rust
use std::ptr;
use std::time::Instant;

fn main() {
    thread_local!(static PTR: usize = 0);

    const ITERS: u32 = 100_000_000;

    let t0 = Instant::now();
    for _ in 0..ITERS {
        PTR.with(|ptr| unsafe { ptr::write_volatile(ptr as *const _ as *mut _, 0) });
    }
    let dur = t0.elapsed();
    println!("elapsed: {:?}, time/iter: {:?}",
             dur,
             dur.checked_div(ITERS).unwrap());
}
```

**Comment by alexcrichton** [comment]
Hm ok, can you gist what the IR looks like for the previous version of this patch? I'm curious about a test case that [looks like this](https://play.rust-lang.org/?gist=f0c49ce1db9d6e6c0f9e0eb545f87bbf&version=stable&mode=release) to see where the perf impact comes into play.

**Comment by joshlf** [comment]
Do you want the IR for the previous version of this patch or rather the current code on nightly? I only ask because I blew away the old version, so I'd have to re-write it from scratch.

Just so I understand what you're trying to figure out - what's the concern with storing multiple callbacks? My reasoning is that only the performance of the `__get` callback matters because it's the only one that's in the fast path. Do you think that the mere fact of storing many callbacks will cause LLVM to not make certain optimizations to the fast path?

**Comment by alexcrichton** [comment]
Ah sorry I thought it may still be around, but yeah I'm curious about the previous approach which didn't increase the number of callbacks and why it was too slow.

I'd be slightly worried about the cross-crate performance here as the fast path may require more than one callback to be called and unfortunately they can't be inlined right now :(

**Comment by joshlf** [comment]
Hmm on second thought, I might be able to sift through the orphaned commits and find it - unless I ran `git gc`, which I do a little too often I'll admit.

The old code was basically: `get` would return a `&LocalKeyState<T>` (recall from my previous comment that `LocalKeyState<T>` is an enum of all four states where one is `Initialized(T)`), and `try_with` would call `get` and match on the return value.

My completely untested suspicion is that the performance issue was due to the fact that the match was over four variants rather than two. Based on that suspicion, I changed to the current version, where the fast path uses an `Option`.

**Comment by joshlf** [comment]
Yeah, no luck - no orphaned commits with the old code.

**Comment by joshlf** [comment]
Any thoughts on this? I'm happy to do more work if you need, but I'd like to get this merged if possible because there are a number of changes in my own work that are blocked on it.

**Comment by alexcrichton** [comment]
Sorry about the delay. I'm personally wary to r+ this without understanding why so many callbacks are necessary, or without understanding the performance impact on cross-crate calls. It would be great if there were an alternate version of this branch to investigate the performance impact of, but otherwise I just haven't had time to experiment myself just yet.

**Comment by joshlf** [comment]
How would you feel about having two callbacks - the existing `get`, and then a single `transition` that accepted an enum identifying the different kinds of transitions that can happen (with each transition implemented as the various callbacks are right now)?

**Comment by alexcrichton** [comment]
That seems plausible, yes, but I still don't understand why "return to me the memory location with storage" doesn't work

**Comment by joshlf** [comment]
OK, I'm working on a rewrite, but I've discovered a serious bug in my implementation (that plagues the current version as well as the new version I'm writing) - the approach that the `mod os` version (that uses OS-provided TLS on platforms without `#[thread_local]`) takes to determining whether TLS is currently being initialized is buggy, and I'm not sure how to work around it.

In particular, I misunderstood the semantics of when `std::sys_common::thread_local::StaticKey::{get, set}` allocate. I'm concerned that there may actually be no good way of detecting whether an allocation routine is currently running.

I have two thoughts: on the one hand, maybe I can figure out a way to detect this. On the other hand, if I can't, I may have to give up and just document that allocator implementations are responsible for handling this themselves.

**Comment by joshlf** [comment]
Hey @alexcrichton , can I ask you a question about some code that `git blame` says you wrote? In particular, this destructor for the OS implementation of TLS:

```rust
unsafe extern fn destroy_value<T: 'static>(ptr: *mut u8) {
        // The OS TLS ensures that this key contains a NULL value when this
        // destructor starts to run. We set it back to a sentinel value of 1 to
        // ensure that any future calls to `get` for this thread will return
        // `None`.
        //
        // Note that to prevent an infinite loop we reset it back to null right
        // before we return from the destructor ourselves.
        let ptr = Box::from_raw(ptr as *mut Value<T>);
        let key = ptr.key;
        key.os.set(1 as *mut u8);
        drop(ptr);
        key.os.set(ptr::null_mut());
    }
```

When you set the pointer back to null before returning, doesn't that make it so that a future access of the key after the destructor has run will cause the key to be re-initialized because a null pointer looks like an uninitialized key rather than a destroyed one? This could happen if the destructor of another TLS key accesses this key, and that destructor is called after this key's destructor.

**Comment by eddyb** [comment]
@joshlf It's tri-state, IIRC they go like this (conceptually, the actual types might differ):
* uninitialized: `Some(None)`
* initialized: `Some(Some(x))`
* destruction started: `None`

**Comment by joshlf** [comment]
@eddyb I believe you're right, but my concern is what happens after the destruction has finished (after which, as the comment that I quoted says, the pointer is set back from 1 to 0). The implementation of `get` makes it seem that a call to `get` that takes place after the destructor has returned (and thus the pointer is again set to 0) would think that the key was simply not yet initialized, and would thus initialize it again:

```rust
        pub unsafe fn get(&'static self) -> Option<&'static UnsafeCell<Option<T>>> {
            let ptr = self.os.get() as *mut Value<T>;
            if !ptr.is_null() {
                if ptr as usize == 1 {
                    return None
                }
                return Some(&(*ptr).value);
            }

            // If the lookup returned null, we haven't initialized our own
            // local copy, so do that now.
            let ptr: Box<Value<T>> = box Value {
                key: self,
                value: UnsafeCell::new(None),
            };
            let ptr = Box::into_raw(ptr);
            self.os.set(ptr as *mut u8);
            Some(&(*ptr).value)
        }
```

**Comment by alexcrichton** [comment]
@joshlf 

> When you set the pointer back to null before returning, doesn't that make it so that a future access of the key after the destructor has run will cause the key to be re-initialized because a null pointer looks like an uninitialized key rather than a destroyed one? 

Indeed!

We don't have much of an option here, though, unless we track this data elsewhere. If we leave it as 1 then the OS will continue to call the destructor function rather than returning.

**Comment by joshlf** [comment]
OK, in that case I think I'll clarify that in the documentation, and add a TODO in case somebody comes along and figures out how to fix it.

**Comment by joshlf** [comment]
OK, it's ready for review. Just to recap:
- The new version explicitly does not attempt to move into state `Initializing` before performing any allocation (although, as written, it happens to do this for the fast implementation), and there is a warning about using TLS for global allocators in the documentation.
- This version uses two callbacks: `get` and `register_dtor`. The first returns a `&UnsafeCell<LocalKeyValue<T>>`, and the second registers the destructor for the fast implementation (it's a no-op in the OS implementation).

File: src/libstd/thread/local.rs [hunk 1]
```diff
@@ -359,13 +456,25 @@ impl<T: 'static> LocalKey<T> {
...
-            })?;
-            Ok(f(match *slot.get() {
-                Some(ref inner) => inner,
-                None => self.init(slot),
-            }))
+            let slot = (self.get)();
+            if let &LocalKeyValue::Valid(ref inner) = &*slot.get() {
+                // Do this in a separate if branch (rather than just part of the
+                // match statement in the else block) to increase the performance
+                // of the fast path.
```
Comment (alexcrichton) [inline src/libstd/thread/local.rs]: Does this measurably help the codegen?

I'd naively expect not, butI'm curious if you have benchmarks.

File: src/libstd/thread/local.rs [hunk 2]
```diff
@@ -97,10 +96,19 @@ pub struct LocalKey<T: 'static> {
...
+
+    // Platform-specific callbacks.
+
+    // Get a reference to the value. get can be called at any time during the
+    // value's life cycle (Uninitialized, Initializing, Valid, and Destroyed).
+    get: unsafe fn() -> &'static UnsafeCell<LocalKeyValue<T>>,
+
+    // Register a destructor for the value. register_dtor should be called
+    // after the key has been transitioned into the Valid state.
+    register_dtor: unsafe fn(),
```
Comment (alexcrichton) [inline src/libstd/thread/local.rs]: Couldn't this still happen on the first call to `get`? If initialization fails that means we'll just run the destructor on a noop value, right?

File: src/libstd/thread/local.rs [hunk 3]
```diff
@@ -424,36 +529,30 @@ pub mod fast {
...
-        if requires_move_before_drop() {
-            ptr::read((*ptr).inner.get());
-        } else {
-            ptr::drop_in_place((*ptr).inner.get());
-        }
+        let tmp = ptr::read((*ptr).inner.get());
+        // Set inner to Destroyed before we drop tmp so that a
+        // recursive call to get (called from the destructor) will
+        // be able to detect that the value is already being dropped.
+        ptr::write((*ptr).inner.get(), __LocalKeyValue::Destroyed);
```
Comment (alexcrichton) [inline src/libstd/thread/local.rs]: I think `mem::replace` can be used here instead of read/write?

File: src/libstd/thread/local.rs [hunk 4]
```diff
@@ -466,35 +565,35 @@ pub mod os {
...
         }
 
-        pub unsafe fn get(&'static self) -> Option<&'static UnsafeCell<Option<T>>> {
+        pub unsafe fn get(&self) -> &'static UnsafeCell<__LocalKeyValue<T>> {
             let ptr = self.os.get() as *mut Value<T>;
             if !ptr.is_null() {
                 if ptr as usize == 1 {
-                    return None
+                    // The destructor was already called (and set self.os to 1).
+                    return &*(&self.dummy_destroyed as *const _);
```
Comment (alexcrichton) [inline src/libstd/thread/local.rs]: How come `get` no longer takes `'static self` to require these casts?

File: src/libstd/thread/local.rs (hunk 2 above)
Comment (joshlf) [reply src/libstd/thread/local.rs]: I wanted to avoid adding an extra branch in the hot path, and that would require branching in `get` to see whether the destructor had been registered already.

File: src/libstd/thread/local.rs (hunk 3 above)
Comment (joshlf) [reply src/libstd/thread/local.rs]: Good call.

File: src/libstd/thread/local.rs (hunk 4 above)
Comment (joshlf) [reply src/libstd/thread/local.rs]: Oh, honestly I think I wasn't sure what to do there so I just copied what `mod fast` does, but of course that doesn't make sense because `mod os` uses a truly static value. I'll change it back.

File: src/libstd/thread/local.rs (hunk 1 above)
Comment (joshlf) [reply src/libstd/thread/local.rs]: So I'm actually not sure. I tried benchmarking it, but I'm not confident that my `xargo` command is testing what I think it is because I ran it with a version of the code that had a compiler error, and it still ran properly, so I'm not sure what artifact `xargo` is picking up to link as libstd. Here's what I'm running:

`RUST_BACKTRACE=1 RUSTFLAGS='-C opt-level=3' XARGO_RUST_SRC=/path/to/rust/src/ xargo run --release`

Given this command, it looks like there's not much of a difference, but I'm skeptical for the reasons mentioned. If you want to give it a try yourself (since I assume you're better at this whole build pipeline thing), it would be give me more confidence.

**Comment by alexcrichton** [comment]
Hm ok so looking into the perf impact here, I'm getting more and more wary about this change...

Right now the `try_with` function can't be inlined presumably because it's recursive, so there's an unconditional function call on calls to `with` where there isn't today. Using just a bare `match` and fixing the recursion produces an *almost* optimal set of IR, except that now all calls to `with` are unconditionally flagged as "this can panic" because a 'being initialized' stat can get encountered, triggering the panic. 

Today, however, calls to `with` only panic if the closure panics (in the optimal fast case), and it appears everything is otherwise inlined.

In other words, I think that fundamentally adding an "initializing" state is going to come at a perf hit to all thread locals. With that in mind, I wonder if there are perhaps other methods to tackle this for your use case? I've always imagined that `thread_local!` is at a *much* higher level of abstraction that a global allocator because `thread_local!` itself will allocate internaly.

**Comment by joshlf** [comment]
> Today, however, calls to `with` only panic if the closure panics (in the optimal fast case), and it appears everything is otherwise inlined.

Can't `with` already panic (in today's nightly) if the key is in state `Destroyed`? The current source of `with`:

```rust
    pub fn with<F, R>(&'static self, f: F) -> R
                      where F: FnOnce(&T) -> R {
        self.try_with(f).expect("cannot access a TLS value during or \
                                 after it is destroyed")
    }
```

> With that in mind, I wonder if there are perhaps other methods to tackle this for your use case? I've always imagined that `thread_local!` is at a much higher level of abstraction that a global allocator because `thread_local!` itself will allocate internally.

I already decided to give up on making this work for allocation (see [this comment](https://github.com/rust-lang/rust/pull/43550#issuecomment-327676725)). However, this is still useful for other recursive initialization dependencies.

**Comment by alexcrichton** [comment]
I [commented awhile ago](https://github.com/rust-lang/rust/issues/17954#issuecomment-320561197) with [this example](https://play.rust-lang.org/?gist=9616a817659dba2dc5e98c96c96d96ae&version=nightly) which generates [this IR](https://gist.github.com/alexcrichton/add34401bd905c7a3e2489a47a0b373c) for a `FOO.with(|a| *a)` where `a` is just a `Cell<u32>`. There aren't any panics there.

In general yes, `with` can panic, but it requires destructors, I believe, which aren't always set in some cases.

> I already decided to give up on making this work for allocation (see this comment). However, this is still useful for other recursive initialization dependencies.

In that case I'm somewhat inclined to close this? This PR isn't needed for you use case, is technically a regression in behavior, and is currently a regression in performance.

**Comment by joshlf** [comment]
OK, fair enough. I'll open a PR to clarify that TLS constructors cannot depend recursively on the values that they construct, but other than the case of allocators (and perhaps some weird dependency injection scenarios), this should only arise within a given crate, in which case it's more reasonable to ask authors to just architect their stuff differently.

**Comment by joshlf** [comment]
PR added: https://github.com/rust-lang/rust/pull/44396

**Comment by alexcrichton** [comment]
Ok, closing in favor of https://github.com/rust-lang/rust/pull/44396 in that case
